File: python/src/code_smells/change_preventers/parallel_inheritance_hierarchy.py
"""
Parallel Inheritance Hierarchy Code Smell Example

When you create a subclass in one hierarchy, you must create a corresponding
subclass in another hierarchy. The hierarchies mirror each other.

Refactoring suggestion: Move Method and Move Field to eliminate one hierarchy
"""
from abc import ABC, abstractmethod


class Component(ABC):
    """Abstract component hierarchy"""

    @abstractmethod
    def draw(self, renderer: 'Renderer') -> str:
        pass


class Button(Component):
    """Each component type needs a corresponding renderer method"""

    def __init__(self, label: str) -> None:
        self.label = label

    def draw(self, renderer: 'Renderer') -> str:
        return renderer.render_button(self)


class TextBox(Component):
    """Each component type needs a corresponding renderer method"""

    def __init__(self, text: str) -> None:
        self.text = text

    def draw(self, renderer: 'Renderer') -> str:
        return renderer.render_text_box(self)


class Renderer(ABC):
    """Abstract renderer hierarchy mirrors Component hierarchy"""

    @abstractmethod
    def render_button(self, button: Button) -> str:
        pass

    @abstractmethod
    def render_text_box(self, textbox: TextBox) -> str:
        pass


class HtmlRenderer(Renderer):
    """For each Component subclass, we need a render method here"""

    def render_button(self, button: Button) -> str:
        return f"<button>{button.label}</button>"

    def render_text_box(self, textbox: TextBox) -> str:
        return f'<input value="{textbox.text}"/>'


class MarkdownRenderer(Renderer):
    """For each Component subclass, we need a render method here"""

    def render_button(self, button: Button) -> str:
        return f"[{button.label}]"

    def render_text_box(self, textbox: TextBox) -> str:
        return f"_{textbox.text}_"


def demo_parallel_hierarchy() -> list[str]:
    components: list[Component] = [Button('Save'), TextBox('name')]
    renderer = HtmlRenderer()
    return [component.draw(renderer) for component in components]

File: go/code_smells/change_preventers/parallel_inheritance_hierarchy.go
package change_preventers

import "fmt"

// Code smell: Parallel Inheritance Hierarchy [Jerarquía de herencia paralela].
// Agregar un nuevo componente de UI obliga a agregar métodos correspondientes en cada renderer,
// haciendo que ambas jerarquías tengan que crecer al unísono.

// Ejercicio: Añade un componente Image que muestre una imagen.

// Necesitarás añadir Image y renderImage a Renderer, e implementarlo en todos los renderers,
// mostrando cambios en paralelo.

type Renderer interface {
	RenderButton(b *Button) string
	RenderTextBox(t *TextBox) string
}

type Component interface {
	Draw(renderer Renderer) string
}

// Button component
type Button struct {
	Label string
}

func (b *Button) Draw(renderer Renderer) string {
	return renderer.RenderButton(b)
}

// TextBox component
type TextBox struct {
	Text string
}

func (t *TextBox) Draw(renderer Renderer) string {
	return renderer.RenderTextBox(t)
}

// HtmlRenderer implementation
type HtmlRenderer struct{}

func (h *HtmlRenderer) RenderButton(b *Button) string {
	return fmt.Sprintf("<button>%s</button>", b.Label)
}

func (h *HtmlRenderer) RenderTextBox(t *TextBox) string {
	return fmt.Sprintf("<input value=\"%s\"/>", t.Text)
}

// MarkdownRenderer implementation
type MarkdownRenderer struct{}

func (m *MarkdownRenderer) RenderButton(b *Button) string {
	return fmt.Sprintf("[%s]", b.Label)
}

func (m *MarkdownRenderer) RenderTextBox(t *TextBox) string {
	return fmt.Sprintf("_%s_", t.Text)
}

func DemoParallelHierarchy() []string {
	comps := []Component{
		&Button{Label: "Save"},
		&TextBox{Text: "name"},
	}
	renderer := &HtmlRenderer{}
	results := make([]string, len(comps))
	for i, c := range comps {
		results[i] = c.Draw(renderer)
	}
	return results
}
